src/aapns/connection.py

```diff
@@ -0,0 +1,452 @@
+        if now > req.deadline:
+            raise Timeout("Request timed out")
+        if self.closing or self.closed:
+            raise Closed(self.outcome)
+        if self.blocked:
+            raise Blocked()
+
+        try:
+            sid = self.conn.get_next_available_stream_id()
+        except h2.NoAvailableStreamIDError:
+            self.closing = True
+            if not self.outcome:
+                self.outcome = "Exhausted"
+            raise Closed(self.outcome)
+
+        assert sid not in self.channels
+
+        self.last_new_sid = sid
+
+        ch = self.channels[sid] = Channel(sid, None, [])
+        self.conn.send_headers(
+            sid, req.header_with(self.host, self.port), end_stream=False
+        )
+        # FIXME don't we have to increment global flow control
+        # also on reception of stream X data?
+        self.conn.increment_flow_control_window(MAX_RESPONSE_SIZE, stream_id=sid)
+        self.conn.send_data(sid, req.body, end_stream=True)
+        self.please_write.set()
+
+        try:
+            while not self.closed:
+                ch.fut = Future()
+                with suppress(TimeoutError):
+                    await wait_for(ch.fut, req.deadline - now)
+                now = time()
+                if now > req.deadline:
+                    raise Timeout()
+                for event in ch.ev:
+                    if isinstance(event, h2.events.ResponseReceived):
+                        ch.header = dict(event.headers)
+                    elif isinstance(event, h2.events.DataReceived):
+                        ch.body += event.data
+                    elif isinstance(event, h2.events.StreamEnded):
+                        return Response.new(ch.header, ch.body)
+                    elif len(ch.body) >= MAX_RESPONSE_SIZE:
+                        raise ResponseTooLarge()
+            else:
+                raise Closed(self.outcome)
+        finally:
+            # FIXME reset the stream, if:
+            # * connection is still alive, and
+            # * the stream didn't end yet
+            del self.channels[sid]
+
+    async def background_write(self):
+        try:
+            while not self.closed:
+                data = None
+
+                while not data:
+                    if self.closed:
+                        return
+
+                    if data := self.conn.data_to_send():
+                        self.w.write(data)
+                        last_sid = self.last_new_sid
+                        await self.w.drain()
+                        self.last_sent_sid = last_sid
+                    else:
+                        await self.please_write.wait()
+                        self.please_write.clear()
+
+        except ConnectionError as e:
+            if not self.outcome:
+                self.outcome = str(e)
+        except Exception:
+            logger.exception("background write task died")
+        finally:
+            self.closing = self.closed = True
```

Comment:
shouldn't this also clear all the active  channels? like in https://github.com/HENNGE/aapns/blob/87de528690ab4380ba1458d7f165c383a1319bad/src/aapns/connection.py#L273-L275

Reply:
Let me check that.
It could be that I'm relying on the background task termination to do that 🤔 
Will check, thanks!

Reply:
OK, here's my thinking, I hope the code actually worked like this...
When a connection is getting closed, both reader and writer ought to die.
However, there can still be data in the pipe, thus, writer is supposed to exit first, then reader may still read something out, then reader exists, at which point all pending requests must wake up.

Thus:
* writer exit: no wakeup
* reader exit: yes wakeup

Reply:
what if the writer dies but the reader does not? I guess that's handled with `closing`/`closed` somehow?

Reply:
yep, if anything dies, then closing/closed is set and the rest of tasks will die too.
(which means, probably data in the pipe will not be read out, sigh!)